os/net/ipv6/uip-ds6-nbr.c

```diff
@@ -55,28 +55,85 @@
 #include "net/ipv6/uip-nd6.h"
 #include "net/routing/routing.h"
 
+#if UIP_DS6_NBR_MULTI_IPV6_ADDRS
+#include "lib/memb.h"
+#endif /* UIP_DS6_NBR_MULTI_IPV6_ADDRS */
+
 /* Log configuration */
 #include "sys/log.h"
 #define LOG_MODULE "IPv6 Nbr"
 #define LOG_LEVEL LOG_LEVEL_IPV6
 
+#if UIP_DS6_NBR_MULTI_IPV6_ADDRS
+static void add_uip_ds6_nbr_to_nbr_entry(uip_ds6_nbr_t *nbr,
+                                         uip_ds6_nbr_entry_t *nbr_entry);
+static void remove_uip_ds6_nbr_from_nbr_entry(uip_ds6_nbr_t *nbr);
+static void remove_nbr_entry(uip_ds6_nbr_entry_t *nbr_entry);
+static void free_uip_ds6_nbr(uip_ds6_nbr_t *nbr);
+static void uip_ds6_nbr_entry_rm(uip_ds6_nbr_entry_t *nbr_entry);
+NBR_TABLE(uip_ds6_nbr_entry_t, uip_ds6_nbr_entries);
+MEMB(uip_ds6_nbr_memb, uip_ds6_nbr_t, UIP_DS6_NBR_MAX_NEIGHBOR_CACHES);
+#else
 NBR_TABLE(uip_ds6_nbr_t, ds6_neighbors);
+#endif /* UIP_DS6_NBR_MULTI_IPV6_ADDRS */
 
 /*---------------------------------------------------------------------------*/
 void
 uip_ds6_neighbors_init(void)
 {
   link_stats_init();
+#if UIP_DS6_NBR_MULTI_IPV6_ADDRS
+  memb_init(&uip_ds6_nbr_memb);
+  nbr_table_register(uip_ds6_nbr_entries,
+                     (nbr_table_callback *)uip_ds6_nbr_entry_rm);
+#else
   nbr_table_register(ds6_neighbors, (nbr_table_callback *)uip_ds6_nbr_rm);
+#endif /* UIP_DS6_NBR_MULTI_IPV6_ADDRS */
 }
 /*---------------------------------------------------------------------------*/
 uip_ds6_nbr_t *
 uip_ds6_nbr_add(const uip_ipaddr_t *ipaddr, const uip_lladdr_t *lladdr,
                 uint8_t isrouter, uint8_t state, nbr_table_reason_t reason,
                 void *data)
 {
-  uip_ds6_nbr_t *nbr = nbr_table_add_lladdr(ds6_neighbors, (linkaddr_t*)lladdr
-                                            , reason, data);
+  uip_ds6_nbr_t *nbr;
+
+#if UIP_DS6_NBR_MULTI_IPV6_ADDRS
+  uip_ds6_nbr_entry_t *nbr_entry;
+
+  assert(uip_ds6_nbr_lookup(ipaddr) == NULL);
+  if(uip_ds6_nbr_lookup(ipaddr)) {
+    LOG_ERR("%s: uip_ds6_nbr for ", __func__);
+    LOG_ERR_6ADDR(ipaddr);
+    LOG_ERR_("has already existed\n");
+    return NULL;
+  }
+
+  nbr_entry = nbr_table_get_from_lladdr(uip_ds6_nbr_entries,
+                                        (const linkaddr_t *)lladdr);
+  if(nbr_entry == NULL) {
+    if((nbr_entry =
+        nbr_table_add_lladdr(uip_ds6_nbr_entries,
+                             (linkaddr_t*)lladdr, reason, data)) == NULL) {
+      LOG_ERR("%s: cannot allocate a new uip_ds6_nbr_entry\n", __func__);
+      return NULL;
+    } else {
+      LIST_STRUCT_INIT(nbr_entry, uip_ds6_nbrs);
+    }
+  }
+
+  if((nbr = (uip_ds6_nbr_t *)memb_alloc(&uip_ds6_nbr_memb)) == NULL) {
```

Comment:
Wouldn't it make sense to check if there is free space in this memb *before* doing the nbr_table_add? This would avoid involving the nbr_table pruning to add an entry that will be removed straight away.

Reply:
That makes sense!